llvm/lib/Target/EVM/EVMStackifyCodeEmitter.cpp

```diff
@@ -0,0 +1,574 @@
+//===--- EVMStackifyCodeEmitter.h - Create stackified MIR -------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file transforms MIR to the 'stackified' MIR.
+//
+//===----------------------------------------------------------------------===//
+
+#include "EVMStackifyCodeEmitter.h"
+#include "EVMMachineFunctionInfo.h"
+#include "EVMStackDebug.h"
+#include "EVMStackShuffler.h"
+#include "TargetInfo/EVMTargetInfo.h"
+#include "llvm/MC/MCContext.h"
+
+using namespace llvm;
+
+#define DEBUG_TYPE "evm-stackify-code-emitter"
+
+int EVMStackifyCodeEmitter::CodeEmitter::getStackHeight() const {
+  return StackHeight;
+}
+
+void EVMStackifyCodeEmitter::CodeEmitter::init(MachineBasicBlock *MBB,
+                                               int Height) {
+  StackHeight = Height;
+  CurMBB = MBB;
+  LLVM_DEBUG(dbgs() << "\n"
+                    << "Set stack height: " << StackHeight << "\n");
+  LLVM_DEBUG(dbgs() << "Setting current location to: " << MBB->getNumber()
+                    << "." << MBB->getName() << "\n");
+}
+
+void EVMStackifyCodeEmitter::CodeEmitter::emitInst(const MachineInstr *MI) {
+  unsigned Opc = MI->getOpcode();
+  assert(Opc != EVM::JUMP && Opc != EVM::JUMPI && Opc != EVM::ARGUMENT &&
+         Opc != EVM::RET && Opc != EVM::CONST_I256 && Opc != EVM::COPY_I256 &&
+         Opc != EVM::FCALL && "Unexpected instruction");
+
+  // The stack height is increased by the number of defs and decreased
+  // by the number of inputs. To get the number of inputs, we subtract
+  // the total number of operands from the number of defs, so the
+  // calculation is as follows:
+  // Defs - (Ops - Defs) = 2 * Defs - Ops
+  int StackAdj = (2 * static_cast<int>(MI->getNumExplicitDefs())) -
```

Comment:
1. Consider turning `stackHeight` to `size_t` as there are plenty of operations with `unsigned` values throughout code.
2. Consider making code self-explanatory instead of documenting.
```
size_t NumUses = ...;
StackHeight += NumExplicitDefs;
StackHeight -= NumUses;
```

Reply:
Done and also added assert to check whether there are enough operands on stack.